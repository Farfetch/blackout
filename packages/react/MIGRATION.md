This guide will help you migrate `@farfetch/blackout-react` to the latest version.

# Table of Contents

- [Migrating from react v0 to v1](#migrating-from-react-v0-to-v1)

  - [Analytics](#analytics)
  - [Content](#content)
  - [Locale](#locale)

## Migrating from react v0 to v1

If you are migrating from `@farfetch/blackout-react` v0, there are a few setup steps that you will need to follow first which are detailed below. Later in this guide, there are sections that contains the migration steps for each specific area so you can find the changes you need more easily.

### Install `@farfetch/blackout-react` v1 and its peer dependencies

You will need to install the `@farfetch/blackout-react` package v1 to use the modules you would import from `@farfetch/blackout-react` v0.

Follow the [installation](README.md#installation) instructions to install the package and its peer dependencies.

### Change bundler/jest settings to support ESM module format

All modules provided by the new `@farfetch/blackout-react` v1 package are in ESM-only format now. Check the [configuration](README.md#configuration) section to know what changes you might need to do to use this module format in your applications.

### Remove any alias to `src` folders

The packages do not include the `src` folder under its root anymore so any alias configurations you might have in your bundler/jest/typescript configurations are not needed now as they include the areas' folders directly under its root.

### Use a typescript-enabled IDE

The `@farfetch/blackout-react` v1 package is entirely authored in typescript and ships with types that when used in conjunction with a typescript-enabled IDE (like VSCode) can help with the migration by suggesting imports and giving errors for invalid usage of the modules. Make sure you use an IDE that supports typescript to make the migration process easier.

### Change import style

The package supports 2 import styles: You can import everything from the root of the package or you can import directly from a file. We recommend using imports from the root of the package since we do not guarantee that imports directly to a file will be kept in future versions.

```js
// Previously (Still works but should be used only when it is not possible to import from the root of the package):
import { useBagItem } from '@farfetch/blackout-react/bags';

// Change to:
import { useBagItem } from '@farfetch/blackout-react';
```

Tip: Use VSCode's import suggestions to help you import the modules you need correctly.

### Hooks interface changes

All hooks implementations have changed their interfaces in order to be more structured. Hooks that can auto fetch data now receive an `enableAutoFetch` boolean option parameter so you can control this feature. Output from hooks also changed to have the data they return under the `data` property and any actions under the `actions` property of the output object. Typescript will point out any errors and will list all available properties you can use in both input and output values.

### Analytics

#### Renamed exports

The following table contains the exports that were renamed and their new names you must use now.

| Old name                   | New name                 |
| -------------------------- | ------------------------ |
| analytics (default export) | analytics (named export) |
| eventTypes                 | EventType                |
| fromParameterTypes         | FromParameterType        |
| interactionTypes           | InteractionType          |
| loginMethodParameterTypes  | LoginMethodParameterType |
| pageTypes                  | PageType                 |
| platformTypes              | PlatformType             |

#### Removed exports

The following table contains the exports that were removed and the alternatives you can use, if any.

| Removed export                        | Notes                                                                     |
| ------------------------------------- | ------------------------------------------------------------------------- |
| integrations.AnalyticsService         | Removed as it was deprecated                                              |
| integrations.Nethone                  | Removed as it was deprecated                                              |
| integrations.Vitorino                 | Use `Forter` and `Riskified` integrations directly                        |
| integrations.Zaraz                    | Removed as it was deprecated                                              |
| CLIENT_ID_HEADER_NAME (from CastleV2) | Changed to a static member with the same name on the `Castle` integration |
| CASTLE_MESSAGE_PREFIX (from CastleV2) | Changed to a static member with the same name on the `Castle` integration |

### Content

#### Renamed exports

The following table contains the exports that were renamed and their new names you must use now.

| Old name                | New name                 |
| ----------------------- | ------------------------ |
| generateArticle         | structuredArticle        |
| generateBreadcrumbsList | structuredBreadcrumbs    |
| generateOrganization    | structuredOrganization   |
| generateProduct         | structuredProduct        |
| generateProductListing  | structuredProductListing |
| generateVideoObject     | structuredVideoObject    |
| generateWebsiteSearch   | structuredWebsiteSearch  |
| useCommercePage         | useCommercePages         |
| useContentType          | useContents              |
| usePage                 | useContentPage           |
| useSEO                  | useSeoMetadata           |

#### Removed exports

The following table contains the exports that were removed and the alternatives you can use, if any.

| Removed export  | Notes                                                                                                                                                            |
| --------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Article         | Use `structuredArticle` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet`        |
| Breadcrumbs     | Use `structuredBreadcrumbs` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet`    |
| Head            | Use `useSeoMetadata` hook to obtain seo metadata. You will need to set the metadata yourself in the response with a solution like `react-helmet`                 |
| Product         | Use `structuredProduct` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet`        |
| ProductsListing | Use `structuredProductListing` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet` |
| Organization    | Use `structuredOrganization` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet`   |
| VideoObject     | Use `structuredVideoObject` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet`    |
| WebsiteSearch   | Use `structuredWebsiteSearch` function to obtain the script tag. You will need to set the metadata yourself in the response with a solution like `react-helmet`  |
| useNavbars      | Use `useContents` hook instead by specifying the `contentTypeCode` param as `navbars` and the `codes` param with the navigation key you want to fetch            |
| useWidget       | Use `useContents` hook instead by specifying the `contentTypeCode` param as `widgets` and the `codes` param with the widget key you want to fetch                |
| useAction       | Create your own hook that performs the same logic as this hook is only to be used internally                                                                     |

#### Other notable changes

##### react-helmet dependency was removed

In the previous version, there was an explicit dependency between this package and `react-helmet` as it was being used by some exports (example: `Head` and `Article` components) to set the document's head with metadata. In this version, that coupling was removed as well as the components that used it. Now you have total control to use your own solution to set metadata information and this package provides helpers that you can use to help you with that.

### Locale

#### New hooks to deal with locale structures

The following table shows the hooks that were added that can be used to deal with locale structures:

| Hook                     | Notes                                                                                                                                                      |
| ------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
| useCountries             | Use `useCountries` to get all countries.                                                                                                                   |
| useCountryAddressSchemas | Use `useCountryAddressSchemas` to get all country address schemas by country code.                                                                         |
| useCountryStateCities    | Use `useCountryStateCities` to get all cities by country code and state id.                                                                                |
| useCountryStates         | Use `useCountryStates` to get all states by country code.                                                                                                  |
| useLocale                | Use `useLocale` to get all info of locale(`subfolder`, `countryCultureCode`, `countryCode`, `country`, `sourceCountryCode` and `currency`) from selectors. |

#### `currencyFormatter` util

A new utility called `currencyFormatter` was added that can be used to format currency-based values.
