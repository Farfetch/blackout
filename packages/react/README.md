# @farfetch/blackout-react

React components, hooks and other tools filled with business logic to help using the Farfetch Platform Solutions' services in web or native e-commerce apps.

## Installation

**yarn**

```sh
yarn add @farfetch/blackout-react
```

**npm**

```sh
npm i @farfetch/blackout-react
```

### Peer dependencies

Make sure that you have installed the correct peer dependencies of this package:

- [`@farfetch/blackout-client`](https://www.npmjs.com/package/@farfetch/blackout-client)
- [`@farfetch/blackout-redux`](https://www.npmjs.com/package/@farfetch/blackout-redux)
- [`axios`](https://www.npmjs.com/package/axios)
- [`lodash`](https://www.npmjs.com/package/lodash)
- [`react`](https://www.npmjs.com/package/react)
- [`react-redux`](https://www.npmjs.com/package/react-redux)
- [`redux`](https://www.npmjs.com/package/redux)

## Usage

You just need to import and use what you need

```js
// Managing a bag item
import { useBagItem } from '@farfetch/blackout-react/bags';

const { error, isLoading } = useBagItem(bagItemId);
```

### Additional configuration

Since this package is published in its original structure, all the source code is contained in a `src` folder. This means you might need additional configurations:

- In order to have friendly imports (`@farfetch/blackout-react` vs `@farfetch/blackout-react/src`), you probably want to add aliases

  ```js
  // Webpack example
  config.resolve.alias = {
    '@farfetch/blackout-react': '@farfetch/blackout-react/src',
  };
  ```

- In order to have your project running, you probably need a specific loader
  ```js
  // Webpack example
  config.module.rules.push({
    test: /\.jsx?$/,
    include: [/node_modules\/@farfetch\/blackout-react/],
    use: [
      {
        loader: 'babel-loader',
        options: myBabelConfig,
      },
    ],
  });
  ```

## Contributing

Pull requests are welcome! For major changes, please open an issue first to discuss what you would like to change.

Please read the [CONTRIBUTING](../../CONTRIBUTING.md) file to know what we expect from your contribution and the guidelines you should follow.

## License

[MIT](../../LICENSE) @ Farfetch
