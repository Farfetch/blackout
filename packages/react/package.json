{
  "name": "@farfetch/blackout-react",
  "version": "1.0.0-next.238",
  "description": "React components, hooks and other tools filled with business logic to help using the Farfetch Platform Solutions' services in web or native e-commerce apps",
  "license": "MIT",
  "exports": {
    "./*": "./dist/*"
  },
  "files": [
    "dist",
    "src"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/Farfetch/blackout.git"
  },
  "dependencies": {
    "@castleio/castle-js": "^2.1.13",
    "@farfetch/blackout-analytics": "^1.0.0-next.21",
    "crypto-js": "3.1.9-1",
    "prop-types": "^15.7.2",
    "proper-url-join": "^2.1.1",
    "react-iframe": "^1.8.0",
    "url-parse": "^1.4.4",
    "yup": "^0.32.11"
  },
  "devDependencies": {
    "@farfetch/blackout-analytics": "^0.1.0-next.17",
    "@farfetch/blackout-client": "^2.0.0-next.156",
    "@farfetch/blackout-redux": "^1.0.0-next.242",
    "axios": "0.21.4"
  },
  "peerDependencies": {
    "@farfetch/blackout-analytics": "^0.1.0-next.17",
    "@farfetch/blackout-client": "^2.0.0-next.139",
    "@farfetch/blackout-redux": "^1.0.0-next.217",
    "lodash": "^4.17.21",
    "react": "^16.7.0",
    "react-redux": "^7.2.0",
    "redux": "^4.1.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
