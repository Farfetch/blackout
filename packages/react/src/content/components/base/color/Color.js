import { colorStyle } from './styles';
import PropTypes from 'prop-types';
import React from 'react';

const Color = ({ data: { hex, displayOptions } }) => (
  <div>
    {displayOptions?.displayName && (
      <span>{`${displayOptions?.displayName} (${hex})`}</span>
    )}
    <div style={colorStyle(hex)} />
  </div>
);

Color.propTypes = {
  data: PropTypes.object.isRequired,
};

export default Color;
