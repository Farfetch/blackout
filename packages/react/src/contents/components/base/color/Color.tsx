import React from 'react';
import type { ColorComponent } from '../../../types/index.js';

const Color = ({ data: { hex } }: ColorComponent): JSX.Element | null =>
  hex ? <div>{hex}</div> : null;

export default Color;
