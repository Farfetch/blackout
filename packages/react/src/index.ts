/**
 * Blackout React.
 */
export * from './analytics';
export * from './authentication';
export * from './bags';
export * from './categories';
export * from './checkout';
export * from './helpers';
export * from './products';
export * from './search';
export * from './users';
export * from './wishlists';
