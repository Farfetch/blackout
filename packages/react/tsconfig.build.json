{
  "extends": "../../configs/typescript/tsconfig.package-build.json",
  "compilerOptions": {
    "rootDir": "src"
  },
  "exclude": ["**/*.test.ts*", "**/*.types.ts", "**/__fixtures__/*"],
  "include": ["src"]
}
