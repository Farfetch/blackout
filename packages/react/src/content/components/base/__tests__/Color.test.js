import { render } from '@testing-library/react';
import Color from '../color/Color';
import React from 'react';

describe('<Color />', () => {
  const data = {
    displayOptions: {
      displayName: 'Color',
    },
    hex: '#000',
  };

  it('should render properly', () => {
    const { container } = render(<Color data={data} />);

    expect(container).toMatchSnapshot();
  });

  it('should not render text if it is not a display name', () => {
    const { queryByText } = render(<Color data={data} />);

    expect(queryByText('Color')).not.toBeInTheDocument();
  });
});
