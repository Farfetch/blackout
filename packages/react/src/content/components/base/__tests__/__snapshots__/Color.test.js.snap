// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Color /> should render properly 1`] = `
<div>
  <div>
    <span>
      Color (#000)
    </span>
    <div
      style="width: 20px; height: 20px; background-color: rgb(0, 0, 0);"
    />
  </div>
</div>
`;
